libbeat/statestore/backend/memlog/diskstore.go

```diff
@@ -165,5 +264,410 @@ func (s *diskstore) LogOperation(op op) error {
+	})
+
+	// delete old transaction files
+	active, _ := activeDataFile(s.dataFiles)
+	updateActiveMarker(s.log, s.home, active)
+	s.removeOldDataFiles()
+
+	trySyncPath(s.home)
+	return nil
+}
+
+func (s *diskstore) checkpointTmpFile(baseName string, states map[string]entry) (string, error) {
+	tempfile := baseName + ".new"
+	f, err := os.OpenFile(tempfile, os.O_RDWR|os.O_CREATE|os.O_TRUNC|os.O_SYNC, s.fileMode)
+	if err != nil {
+		return "", err
+	}
+
+	ok := false
+	defer cleanup.IfNot(&ok, func() {
+		f.Close()
+	})
+
+	writer := bufio.NewWriterSize(&ensureWriter{f}, s.bufferSize)
+	enc := newJSONEncoder(writer)
+	if _, err = writer.Write([]byte{'['}); err != nil {
+		return "", err
+	}
+
+	first := true
+	for key, entry := range states {
+		prefix := []byte(",\n")
+		if first {
+			prefix = prefix[1:]
+			first = false
+		}
+		if _, err = writer.Write(prefix); err != nil {
+			return "", err
+		}
+
+		err = enc.Encode(storeEntry{
+			Key:    key,
+			Fields: entry.value,
+		})
+		if err != nil {
+			return "", err
+		}
+	}
+
+	if _, err = writer.Write([]byte("\n]")); err != nil {
+		return "", err
+	}
+
+	if err = writer.Flush(); err != nil {
+		return "", err
+	}
+
+	if err = syncFile(f); err != nil {
+		return "", err
+	}
+
+	ok = true
+	if err = f.Close(); err != nil {
+		return "", err
+	}
+
+	return tempfile, nil
+}
+
+func (s *diskstore) checkpointClearLog() {
+	if s.logFile == nil {
+		s.logNeedsTruncate = true
+		return
+	}
+
+	err := s.logFile.Truncate(0)
+	if err == nil {
+		_, err = s.logFile.Seek(0, os.SEEK_SET)
+	}
+
+	if err != nil {
+		s.logFile.Close()
+		s.logFile = nil
+		s.logBuf = nil
+		s.logNeedsTruncate = true
+		s.logInvalid = true
+	}
+
+	s.logEntries = 0
+	s.logFileSize = 0
+}
+
+func updateActiveMarker(log *logp.Logger, home, active string) error {
+	activeLink := filepath.Join(home, "active.dat")
+	tmpLink := filepath.Join(home, "active.dat")
+	log = log.With("temporary", tmpLink, "data_file", active, "link_file", activeLink)
+
+	if active == "" {
+		if err := os.Remove(activeLink); err != nil { // try, remove active symlink if present.
+			log.Errorf("Failed to remove old pointer file: %v", err)
+		}
+		return nil
+	}
+
+	// Atomically try to update the pointer file to the most recent data file.
+	// We 'simulate' the atomic update by create the temporary active.json.tmp symlink file,
+	// which we rename to active.json. If active.json.tmp exists we remove it.
+	if err := os.Remove(tmpLink); err != nil && !os.IsNotExist(err) {
+		log.Errorf("Failed to remove old temporary symlink file: %v", err)
+		return err
+	}
+	if err := ioutil.WriteFile(tmpLink, []byte(active), 0600); err != nil {
+		log.Errorf("Failed to write temporary pointer file: %v", err)
+		return err
+	}
+	if err := os.Rename(tmpLink, activeLink); err != nil {
+		log.Errorf("Failed to replace link file: %v", err)
+		return err
+	}
+
+	trySyncPath(home)
+	return nil
+}
+
+// removeOldDataFiles sorts the data files by their update sequence number and
```

Comment:
The comment says this sorts the data files, but I don't see any sorting happen. It looks more like it expects an already-sorted data files list. In fact it looks like `diskstore.dataFiles` is _always_ required to be sorted by txid, but I'm not sure of that -- can the definition of `dataFiles` document whether / when it is required to be sorted?

Reply:
The sort happens in `listDataFiles`. We keep the sequence ordered by always trying to delete all but the last entry, and appending new entries to the end. Will add a comment to the `diskstore` type `dataFiles` field that it requires the list to be sorted by ID via `isTxIDLessEqual`